pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
 			continue
 		}
 		if commonConditionMatch(pctx, cp) && clusterMatch(c, cp, hosts) {
+			// backup the Cluster in case mergeTransportSocketCluster modifies it
+			cpValueBkp := proto.Clone(cp.Value)
 
-			ret, err := mergeTransportSocketCluster(c, cp)
+			err := mergeTransportSocketCluster(c, cp)
 			if err != nil {
 				log.Debugf("Merge of transport socket failed for cluster: %v", err)
 				continue
 			}
 			applied = true
-			if !ret {
-				merge.Merge(c, cp.Value)
-			}
+			merge.Merge(c, cp.Value)
+
+			// restore untouched copy of the Cluster for use in next iteration
+			cp.Value = cpValueBkp
 		}
 		IncrementEnvoyFilterMetric(cp.Key(), Cluster, applied)
 	}
 	return c
 }
 
-// Test if the patch contains a config for TransportSocket
-// Returns a boolean indicating if the merge was handled by this function; if false, it should still be called
-// outside of this function.
-func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfigPatchWrapper) (merged bool, err error) {
+func patchTransportSocket(ts *core.TransportSocket, patchTS *core.TransportSocket) (err error) {
+	dstCluster := ts.GetTypedConfig()
+	srcPatch := patchTS.GetTypedConfig()
+
+	if dstCluster != nil && srcPatch != nil {
+
+		retVal, errMerge := util.MergeAnyWithAny(dstCluster, srcPatch)
+		if errMerge != nil {
+			return fmt.Errorf("function MergeAnyWithAny failed for ApplyClusterMerge: %v", errMerge)
+		}
+
+		// Merge the above result with the whole cluster
+		merge.Merge(dstCluster, retVal)
```

Comment:
hm, I saw we did this in another part of the code, but I don't get it.

Isn't retVal already `dstCluster + srcPatch`? So now we are doing `dstCluster = dstCluster + (dstCluster + srcPatch)` which seems like a NOP?

Reply:
TBH. Me neither, so I decided to leave it as is. I suspect `util.MergeAnyWithAny` is doing some sort of preparation of result to replace and merge.Merge is doing a replacement.